File: apps/go/design-pattern/interpreter/interpreter.go
// Package interpreter demonstrates the Interpreter pattern in Go.
// Expressions form a recursive AST that evaluates against a directory listing.
package interpreter

import (
	"os"
	"path/filepath"
	"strings"
)

// Expression is the abstract expression interface.
type Expression interface {
	Evaluate(dir string) []string
}

// All returns all files in the directory.
type All struct{}

func (a *All) Evaluate(dir string) []string {
	entries, err := os.ReadDir(dir)
	if err != nil {
		return nil
	}
	var result []string
	for _, entry := range entries {
		if !entry.IsDir() {
			result = append(result, entry.Name())
		}
	}
	return result
}

// FileName matches files whose name contains the given pattern.
type FileName struct {
	Pattern string
}

func (f *FileName) Evaluate(dir string) []string {
	all := (&All{}).Evaluate(dir)
	var result []string
	for _, name := range all {
		matched, _ := filepath.Match(f.Pattern, name)
		if matched {
			result = append(result, name)
		}
	}
	return result
}

// Bigger matches files larger than the given size in bytes.
type Bigger struct {
	Size  int64
	Inner Expression
}

func (b *Bigger) Evaluate(dir string) []string {
	candidates := b.Inner.Evaluate(dir)
	var result []string
	for _, name := range candidates {
		info, err := os.Stat(filepath.Join(dir, name))
		if err != nil {
			continue
		}
		if info.Size() > b.Size {
			result = append(result, name)
		}
	}
	return result
}

// Not negates an expression: returns files in All that are NOT in the inner expression.
type Not struct {
	Inner Expression
}

func (n *Not) Evaluate(dir string) []string {
	all := (&All{}).Evaluate(dir)
	excluded := toSet(n.Inner.Evaluate(dir))
	var result []string
	for _, name := range all {
		if !excluded[name] {
			result = append(result, name)
		}
	}
	return result
}

// And returns the intersection of two expressions.
type And struct {
	Left  Expression
	Right Expression
}

func (a *And) Evaluate(dir string) []string {
	left := toSet(a.Left.Evaluate(dir))
	right := a.Right.Evaluate(dir)
	var result []string
	for _, name := range right {
		if left[name] {
			result = append(result, name)
		}
	}
	return result
}

// Or returns the union of two expressions.
type Or struct {
	Left  Expression
	Right Expression
}

func (o *Or) Evaluate(dir string) []string {
	seen := make(map[string]bool)
	var result []string
	for _, name := range o.Left.Evaluate(dir) {
		if !seen[name] {
			seen[name] = true
			result = append(result, name)
		}
	}
	for _, name := range o.Right.Evaluate(dir) {
		if !seen[name] {
			seen[name] = true
			result = append(result, name)
		}
	}
	return result
}

// WithExtension is a convenience: matches files ending with the given extension.
type WithExtension struct {
	Ext string
}

func (w *WithExtension) Evaluate(dir string) []string {
	all := (&All{}).Evaluate(dir)
	var result []string
	ext := w.Ext
	if !strings.HasPrefix(ext, ".") {
		ext = "." + ext
	}
	for _, name := range all {
		if strings.HasSuffix(name, ext) {
			result = append(result, name)
		}
	}
	return result
}

func toSet(items []string) map[string]bool {
	s := make(map[string]bool)
	for _, item := range items {
		s[item] = true
	}
	return s
}

File: apps/php/design-pattern/src/Interpreter/Interpreter.php
<?php

declare(strict_types=1);

namespace DesignPattern\Interpreter;

abstract class Expression
{
    abstract public function evaluate(string $dir): array;
}

class All extends Expression
{
    public function evaluate(string $dir): array
    {
        if (!is_dir($dir)) {
            return [];
        }
        $files = [];
        $entries = scandir($dir);
        foreach ($entries as $entry) {
            if ($entry === '.' || $entry === '..') {
                continue;
            }
            $path = $dir . DIRECTORY_SEPARATOR . $entry;
            if (is_file($path)) {
                $files[] = $path;
            }
        }
        return $files;
    }
}

class FileName extends Expression
{
    private string $pattern;

    public function __construct(string $pattern)
    {
        $this->pattern = $pattern;
    }

    public function evaluate(string $dir): array
    {
        $all = (new All())->evaluate($dir);
        return array_values(
            array_filter($all, fn(string $path) => fnmatch($this->pattern, basename($path)))
        );
    }
}

class Bigger extends Expression
{
    private int $sizeBytes;

    public function __construct(int $sizeBytes)
    {
        $this->sizeBytes = $sizeBytes;
    }

    public function evaluate(string $dir): array
    {
        $all = (new All())->evaluate($dir);
        return array_values(
            array_filter($all, fn(string $path) => filesize($path) > $this->sizeBytes)
        );
    }
}

class AndExpression extends Expression
{
    private Expression $left;
    private Expression $right;

    public function __construct(Expression $left, Expression $right)
    {
        $this->left = $left;
        $this->right = $right;
    }

    public function evaluate(string $dir): array
    {
        $leftResult = $this->left->evaluate($dir);
        $rightResult = $this->right->evaluate($dir);
        return array_values(array_intersect($leftResult, $rightResult));
    }
}

class OrExpression extends Expression
{
    private Expression $left;
    private Expression $right;

    public function __construct(Expression $left, Expression $right)
    {
        $this->left = $left;
        $this->right = $right;
    }

    public function evaluate(string $dir): array
    {
        $leftResult = $this->left->evaluate($dir);
        $rightResult = $this->right->evaluate($dir);
        return array_values(array_unique(array_merge($leftResult, $rightResult)));
    }
}

class NotExpression extends Expression
{
    private Expression $expression;

    public function __construct(Expression $expression)
    {
        $this->expression = $expression;
    }

    public function evaluate(string $dir): array
    {
        $all = (new All())->evaluate($dir);
        $excluded = $this->expression->evaluate($dir);
        return array_values(array_diff($all, $excluded));
    }
}
